docs: Update advice on safe use of finalizers

cmcaine opened the pull request:
This commit concentrates the advice in the place that users of `finalizer` are most likely to see it and provides examples.

It's unclear to me if any of this is required if you use the `ReentrantLock`, because `_trylock` will disable finalizers anyway, but I couldn't work it out and I figured this PR might summon someone who knows definitively.

cmcaine committed:
Update advice on safe use of finalizers

This commit concentrates the advice in the place that users of
`finalizer` are most likely to see it and provides examples.

It's unclear to me if any of this is required if you use the
`ReentrantLock`, because `_trylock` will disable finalizers anyway, but
I couldn't work it out and I figured this PR might summon someone who
knows definitively.

vtjnash commented on base/gcutils.jl:
```diff
@@ -49,6 +49,14 @@ Register a function `f(x)` to be called when there are no program-accessible ref
+    Julia may interrupt any other code when it calls a finalizer function so you must take special care to avoid concurrency bugs when writing your finalizers.
+    Even if you use no other concurrency features in your program, one finalizer call can interrupt another and cause a concurrency bug if they interact with the same global state.
+
+    If your finalizer interacts with global state (and it probably will), then you must use a synchronisation strategy for that interaction that is safe even if your finalizer is interrupted by another finalizer.
```
sp: synchronization

cmcaine commented on base/gcutils.jl:
````diff
@@ -81,6 +89,92 @@ mutable struct MyMutableStruct
 end
 ```
+
+# Extended help
````
While you're here, can you comment on why `GC.disable_finalizers()` is required, given that `lock(my_lock)` will disable finalizers once it acquires the lock (for the locks included in Base, and presumably also for any correctly implemented locks in packages)?

---

Re: where to put the info:

The docstrings are much more visible at the REPL and in editors with LSP help previews, so if the info is important then it seems like a good idea to make it easy to find?

If we could put a hyperlink in to the relevant manual section then that would be fine, but the REPL obviously doesn't support those.

Anyway, I have a slight preference, but I can move the content back into the manual if you or other reviewers have stronger preferences for that way.

---

Re: strategy 2, sure, removed. Distributed.jl still uses basically that strategy, tho. Is that a problem?

```julia
function finalize_ref(r::AbstractRemoteRef)
    if r.where > 0 # Handle the case of the finalizer having been called manually
        if trylock(client_refs.lock) # trylock doesn't call wait which causes yields
            try
                delete!(client_refs.ht, r) # direct removal avoiding locks
                if isa(r, RemoteChannel)
                    send_del_client_no_lock(r)
                else
                    # send_del_client only if the reference has not been set
                    v_cache = @atomic :monotonic r.v
                    v_cache === nothing && send_del_client_no_lock(r)
                    @atomic :monotonic r.v = nothing
                end
                r.where = 0
            finally
                unlock(client_refs.lock)
            end
        else
            finalizer(finalize_ref, r)
            return nothing
        end
    end
    nothing
end
```

vtjnash replied:
True, strategy 1 seems to already be implicit in the use of locks (and is mentioned as such), which makes it an odd case to describe as such, instead of just pointing out that locks are always needed

cmcaine committed:
Text revisions suggested by vtjnash

cmcaine committed:
Remove broken crossref

I still don't understand how @ref is supposed to work, apparently.

vchuravy commented:
Thank you for improving the docs. I will try to give a proper review later.

I think we should not give any recommendations that assumes single-threaded execution. Bugs can also arise in a single threaded concurrent execution mode, and we have long planned to move finalizers to a dedicated task and having a utility thread execute that.

cmcaine commented:
Thanks!

The ideal change from my perspective is that we decide that the user should write finalizers in the same way as any other code that could concurrently access shared state. Then the advice can be much more straightforward.

This requires someone deciding that explicit use of `GC.disable_finalizers` is not necessary (because we decide that `lock`/`trylock` should always disable finalizers as they acquire the locks and that that's just as safe as disabling the finalizers before calling `lock`).

vchuravy commented:
Yeah there are two ways forward in making finalizers less special. The one is the change you noted that during a critical region finalizers are disabled. I need to think about this one since we do currently have eager finalizer insertion and there may be a lingering bug there. We may hold the log, the eager finalizer optimization pass inserts the finalizer and the finalizer executes despite us disabling finalizers.


Or (and I personally prefer this) finalizers semantically execute on their own separate task. Thus if a finalizer needs a lock we go and execute the next one. We haven't done this since creating a task per finalizer seems expensive (but we should measure this) and we want finalizers to execute with immediacy and the scheduler may arbitrarily delay then, plus the overhead of task execution.

cmcaine commented:
If an eagerly inserted finalizer can be executed despite finalizers being disabled then I think that bug would equally affect users regardless of whether they called `GC.disable_finalizers() themselves or relied on `lock` disabling finalizers for them. If I'm right about that, then the existence or not of bugs around inserted finalizers doesn't need to be known before a decision is made about what advice to give.

If finalizers may run in semantically independent, uninterruptible tasks in the future then I think the only additional change we should make now is deleting the text about single-threaded strategies, as you suggested above, so I'll do that now.

cmcaine committed:
Simplify

- Removed description of single-threaded solution.
- Assumed that `lock(l)` will always prevent other finalizers
  interrupting us, once it has been acquired (this is not in the
  specification for `AbstractLock`, but all of the locks in `Base` abide
  by it).

cmcaine committed:
Clearly document what locking does to finalizers

cmcaine committed:
Document that 3rd-party locks may not have finalizer guarantee

vtjnash commented:
eager finalizer initialization is supposed to only occur when it does not observe the task state in any way, so it is provably safe to run there, since it cannot detect the difference

cmcaine commented:
Two quesions in bold below.

I tried running some experiments to determine how finalizers interact with locks and `GC.disable_finalizers()`.

If we define some shared state and a lock in task 1, then acquire the lock in task 2, then set up some finalizers in task 1 and exec GC.gc() then our call to GC.gc() blocks (and blocks all waiting finalizers) until task 2 releases the lock.

**Is this what Julia actually does when it does garbage collection on its own
without a manual trigger?** (I think so, from a cursory read of `src/gc.c`, but I don't know).

**Is this desirable?**
The pattern with re-queueing the finalizer when `trylock` fails avoids this.
As does the pattern with enqueueing tasks to be executed later.

I feel like we should recommend a finalizer strategy that will make progress
so long as any finalizer in the batch of running finalizers can make
progress.
(Which is not this)

# Example
```jl
# shared state and lock accessible from the spawned task and finalizers.
my_lock = ReentrantLock()
output = Ref("initial") 

# Spawn a long-running task
Threads.@spawn lock(my_lock) do
    for _ in 1:3
        sleep(1)
        println(".")
    end
    @info "long-running task about to release lock", output[]
end

# Register some finalizer functions
mutable struct M; value; end # a tiny mutable struct to trigger finalization on GC.
for i in 1:3
    finalizer(M(i)) do x
        while !trylock(my_lock)
            @ccall jl_safe_printf("%s"::Cstring, repr(x)::Cstring)::Cvoid
            sleep(.2)
        end
        try
            # Critical section
            output[] = repr(x)
        finally
            unlock(my_lock)
        end
    end;
end

# Trigger the finalizers.
GC.safepoint()
GC.gc()
@info "after GC.gc()", output[]
```

Also I was looking at Gtk.jl because it's called out as an example of the "enqueue and drain later" strategy, and their code does not seem to be concurrency safe, unless I missed something.

https://github.com/JuliaGraphics/Gtk.jl/blob/881fdce8eb5a8bce2f1a846f79ac7142177e5b2a/src/GLib/gtype.jl#L405C1-L422

cmcaine commented on base/gcutils.jl:
```diff
@@ -81,6 +89,57 @@ mutable struct MyMutableStruct
+ 2. Another strategy is to use the finalizer only to schedule the work to be done at another time.
+    This means we don't have to care about finalizers interrupting each other because the actual work will be done in a non-finalizer context.
+
+    Do this by pushing the work into a concurrent queue such as `Base.IntrusiveLinkedListSynchronized{T}`
```
That's fine. I wrote a revision that suggested ConcurrentCollections.jl and also provided an implementation of a bad synchronised queue.

I would like to work out what the broad strokes of the advice should be, which is why I did those experiments and posed those questions in my last posr.

ViralBShah committed:
Update base/gcutils.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>
